dali/python/nvidia/dali/plugin/pytorch.py

```diff
@@ -158,15 +165,19 @@ def __next__(self):
                 pyt_tensors = dict()
                 for category in self._output_categories:
                     pyt_tensors[category] = torch.zeros(category_shapes[category],
-                                                         dtype=category_torch_type[category],
-                                                         device=category_device[category])
+                                                        dtype=category_torch_type[category],
+                                                        device=category_device[category])
 
                 self._data_batches[i][self._current_data_batch] = pyt_tensors
             else:
                 pyt_tensors = self._data_batches[i][self._current_data_batch]
 
             # Copy data from DALI Tensors to torch tensors
             for category, tensor in category_tensors.items():
+                  if self._dynamic_shape and tensor.shape() != list(pyt_tensors[category].size()):
```

Comment:
Just a thought: Do we need the user to provide info about the `dynamic_shape` at all? Without the first part of this `if` condition it would work the same way, right? If shapes are the same no reallocation, if they are not we allocate new memory. The only thing to consider is that it probably is faster to evaluate `bool` and if it is `False` there is no need to compare sizes, but I don't think it is that costly after all. And it would be transparent to the user, so that is a plus, I think.

Reply:
> If shapes are the same no reallocation, if they are not we allocate new memory.

What gives you this guarantee? Can we really trust their allocator to give us the same tensor?

Reply:
Anyway having it be the default behavior is an open question

Reply:
I meant it this is `if tensor.shape() != list(pyt_tensors[category].size()):` only